From bffedbb2f3b228305aa7733ad508f7b3a382827c Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Thu, 9 Aug 2018 19:48:41 +0800
Subject: [PATCH 4319/5242] MGS-4125 [#imx-320] fix 6sx suspend/resume stuck
 issue

commit  94a649f1e7502226b15ce4ead87d73d7b29dec20 from
https://source.codeaurora.org/external/imx/linux-imx.git

6sx gpu need reserve command to flush cache before power-off,
there is no way to perform force sync in power broadcast mode.

command buffer will be exhausted with repeat suspend/resume,
need pull up power to flush command buffer before suspend.

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../hal/os/linux/kernel/gc_hal_kernel_driver.c     |   17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_driver.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_driver.c
index b76b118..4e0c419 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_driver.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_driver.c
@@ -1067,6 +1067,23 @@ static int gpu_suspend(struct platform_device *dev, pm_message_t state)
                 return -1;
             }
 
+            /* need pull up power to flush gpu command buffer before suspend */
+#if gcdENABLE_VG
+            if (i == gcvCORE_VG)
+            {
+                status = gckVGHARDWARE_SetPowerManagementState(device->kernels[i]->vg->hardware, gcvPOWER_ON);
+            }
+            else
+#endif
+            {
+                status = gckHARDWARE_SetPowerManagementState(device->kernels[i]->hardware, gcvPOWER_ON);
+            }
+
+            if (gcmIS_ERROR(status))
+            {
+                return -1;
+            }
+
 #if gcdENABLE_VG
             if (i == gcvCORE_VG)
             {
-- 
1.7.9.5

